src/gha_logs.rs

```diff
@@ -0,0 +1,167 @@
+use crate::github;
+use crate::handlers::Context;
+use anyhow::Context as _;
+use hyper::header::{CONTENT_SECURITY_POLICY, CONTENT_TYPE};
+use hyper::{Body, Response, StatusCode};
+use std::collections::VecDeque;
+use std::str::FromStr;
+use std::sync::Arc;
+use uuid::Uuid;
+
+const ANSI_UP_URL: &str = "https://cdn.jsdelivr.net/npm/ansi_up@6.0.6/+esm";
+const MAX_CACHE_CAPACITY_BYTES: u64 = 50 * 1024 * 1024; // 50 Mb
+
+#[derive(Default)]
+pub struct GitHubActionLogsCache {
+    capacity: u64,
+    entries: VecDeque<(String, Arc<String>)>,
+}
+
+impl GitHubActionLogsCache {
+    pub fn get(&mut self, key: &String) -> Option<&Arc<String>> {
+        if let Some(pos) = self.entries.iter().position(|(k, _)| k == key) {
+            let entry = self.entries.remove(pos).unwrap();
+            self.entries.push_front(entry);
+            Some(&self.entries.front().unwrap().1)
+        } else {
+            None
+        }
+    }
+
+    pub fn put(&mut self, key: String, value: Arc<String>) -> &Arc<String> {
+        let removed = if let Some(pos) = self.entries.iter().position(|(k, _)| k == &key) {
+            self.entries.remove(pos)
+        } else if self.capacity + value.len() as u64 >= MAX_CACHE_CAPACITY_BYTES {
+            self.entries.pop_back()
+        } else {
+            None
+        };
+        if let Some(removed) = removed {
+            self.capacity -= removed.1.len() as u64;
+        }
+        // Insert at the front (most recently used)
+        self.capacity += value.len() as u64;
```

Comment:
Can we check here that we don't store logs exceeding MAX_CACHE_CAPACITY_BYTES? I think as-is we'd be able to store 1 such log (at least until the next request)